WordPress/gutenberg
# Block List: Move deselect behavior to block component

Closes #3712 

This pull request seeks to improve block deselect behavior. The previous implementation had the visual editor test for click events on the canvas to infer a deselection. This was necessary because there are many cases where focus leaves a block, but the block should remain selected (e.g. clicking toolbar buttons, or inspector controls). However, the previous solution was not ideal because (a) it introduces an implicit dependency between the visual editor and block focus, (b) it prioritized click behavior and while tab focus changes _worked_, it was only because multiple singular blocks could not be selected at the same time, and (c) it conflicts with efforts to introduce multiple editor scopes (block nesting).

__Implementation notes:__

I see this as a dramatic simplification to VisualEditor, but still anticipate that improvements could be made. Specifically, the block component shouldn't need to have any awareness of editor headers or sidebars, as the proposed implementation includes. Two alternatives I see here include:

- Relying on [React portal event bubbling behavior](https://reactjs.org/docs/portals.html#event-bubbling-through-portals) to detect changes in focus via events bubbled through virtual ancestry.
   - This requires refactoring toolbar and multi-selection behavior, see #3083.
- Leveraging hooks filtering to make the block component initially unaware of any closest elements, but allow components (such as visual editor) to introduce tests to verify whether the block should be allowed to be deselected

__Testing instructions:__

Verify that there are no regressions in the behavior of selecting blocks, particularly:

- Selecting a block
- Transitioning selection from one block to another, via click and by keyboard tab
- Deselecting a block by clicking outside
- Interacting with a block via toolbar or inspector controls (including toggling between Document and Block settings in sidebar)
